torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -2445,43 +2445,50 @@ Tensor det_backward(const Tensor & grad, const Tensor& self, const Tensor& det)
   };
 
   auto nonsingular_case_backward = [&](const Tensor& grad, const Tensor& self, const Tensor& det) -> Tensor {
-    return unsqueeze_multiple(grad * det, {-1, -2}, self.dim()) * self.inverse().transpose(-2, -1);
+    // Derived from Jacobi's formula for partial derivatives, which can be
+    // found at https://en.wikipedia.org/wiki/Jacobi%27s_formula
+    auto det_grad = (unsqueeze_multiple(det, {-1, -2}, self.dim()) * self.inverse().transpose(-2, -1)).conj();
+    return unsqueeze_multiple(grad, {-1, -2}, self.dim()) * det_grad;
   };
 
   if (self.dim() == 2) {
-    if (det.item<double>() == 0) {
+    if (det.eq(0).item<bool>()) {
       return singular_case_backward(grad, self, det);
     } else {
       return nonsingular_case_backward(grad, self, det);
     }
   } else {
-    auto nonzero_det_indices = at::native::toListOfOptionalTensors(at::where(det));
-    c10::optional<Tensor> first_nonzero_det_index = nonzero_det_indices[0];
+    auto nonzero_det_mask = det.ne(0);
 
-    if (first_nonzero_det_index->size(0) == det.numel()) {  // all determinants are nonzero (non-singular)
+    if (nonzero_det_mask.all().item<bool>()) {
```

Comment:
I wonder if it's actually faster not to special case the all-singular and all-nonsingular cases. I'll check it out

Reply:
Should be faster, right?